internal/docs/design.md

```diff
@@ -184,6 +187,103 @@ needed). However, this would result in inconsistent names over time (e.g., some
 packages would expose `New` with an `Options`, while others would expose
 `NewWithOptions`).
 
+### URLs
+
+To enable the [Backing services factor][] of a Twelve-Factor Application, Go
+Cloud includes the ability to construct each of its API objects using
+identifying URLs. The concrete type's package should include APIs like the
+following:
+
+```go
+package foo
+
+// A type that implements WidgetURLOpener can open widgets based on a URL.
+// The opener must not modify the URL argument. OpenWidgetURL must be safe to
+// call from multiple goroutines.
+//
+// For maximum flexibility, WidgetURLOpeners should not assume that the URL
+// has a particular scheme.
+type WidgetURLOpener interface {
+  OpenWidgetURL(ctx context.Context, u *url.URL) (*Widget, error)
+}
+
+// A type that implements NamedWidgetURLOpener has a preferred scheme for
+// the URLs it opens. WidgetURLScheme must be safe to call from multiple
+// goroutines.
+type NamedWidgetURLOpener interface {
+  WidgetURLOpener
+  WidgetURLScheme() string
+}
+
+// URLMux is a URL opener multiplexer. It matches the scheme of the URLs
+// against a set of registered schemes and calls the opener that matches the
+// URL's scheme.
+//
+// The zero value is a URLMux with no registered schemes.
+type URLMux struct {
+  // ...
+}
+
+// AddWidget registers the opener with its preferred scheme. If an opener
+// already exists for the scheme, AddWidget panics.
```

Comment:
WDYT about allowing re-registrations for the same scheme instead of panic? (last one wins).

This would allow a user to create a custom `URLOpener` for a specific provider and re-register it on the default mux (instead of having to create their own mux).

I'm imagining a scenario where user has multiple providers installed, and is happy with everything but wants to change one default for one provider. Without re-registration, they are forced to create a new mux and register all of the existing providers, then pass the new mux around everywhere -- a big code change for a small semantic change.

Reply:
Panicking is the more reasonable behavior. The library can't distinguish from the usage you describe and two packages that have a naming collision. Package initialization order is well-defined wrt dependencies, but not well-defined across an arbitrary program. Seemingly random restructurings of the program can affect which one gets registered.

(See golang/go#20478 for a related issue of where this can go wrong.)

Reply:
We could have a separate method, `Reregister` or some such, that would allow it; package initializers would never use it but users could.